tasks/change-granule-collection-s3/src/index.ts

```diff
@@ -0,0 +1,388 @@
+'use strict';
+
+import { Context } from 'aws-lambda';
+import { runCumulusTask } from '@cumulus/cumulus-message-adapter-js';
+import get from 'lodash/get';
+import keyBy from 'lodash/keyBy';
+import cloneDeep from 'lodash/cloneDeep';
+import zip from 'lodash/zip';
+// eslint-disable-next-line lodash/import-scope
+import { Dictionary } from 'lodash';
+import path from 'path';
+import pMap from 'p-map';
+import { MissingS3FileError, InvalidArgument } from '@cumulus/errors';
+import { S3 } from '@cumulus/aws-client';
+import {
+  unversionFilename,
+} from '@cumulus/ingest/granule';
+import {
+  isCMRFile,
+  granulesToCmrFileObjects,
+  metadataObjectFromCMRFile,
+} from '@cumulus/cmrjs';
+import { BucketsConfig } from '@cumulus/common';
+import { urlPathTemplate } from '@cumulus/ingest/url-path-template';
+import { constructCollectionId } from '@cumulus/message/Collections';
+import { getCollection } from '@cumulus/api-client/collections';
+import { log } from '@cumulus/common';
+import { CollectionRecord } from '@cumulus/types';
+import { CumulusMessage } from '@cumulus/types/message';
+import { CollectionFile } from '@cumulus/types';
+import { BucketsConfigObject } from '@cumulus/common/types';
+import { copyObject } from '@cumulus/aws-client/S3';
+import { getGranule } from '@cumulus/api-client/granules';
+import { getRequiredEnvVar } from '@cumulus/common/env';
+import { fetchDistributionBucketMap } from '@cumulus/distribution-utils';
+import { AssertionError } from 'assert';
+import { apiGranuleRecordIsValid, isCMRMetadataFile, updateCmrFileCollections, uploadCMRFile, ValidApiFile, ValidGranuleRecord } from './update_cmr_file_collection';
+
+const MB = 1024 * 1024;
+
+type EventConfig = {
+  targetCollection: {
+    name: string,
+    version: string,
+  }
+  buckets: BucketsConfigObject,
+  s3MultipartChunksizeMb?: number,
+  distribution_endpoint: string,
+  cmrGranuleUrlType: string,
+  invalidBehavior: string,
+};
+
+type ChangeCollectionsS3Event = {
+  config: EventConfig,
+  cumulus_config?: {
+    cumulus_context?: {
+      forceDuplicateOverwrite?: boolean,
+    }
+  },
+  input: {
+    granuleIds: Array<string>,
+  }
+};
+
+function getConcurrency() {
+  return Number(process.env.concurrency || 100);
+}
+
+/**
+ * Is this move a "real" move, or is target location identical to source
+ */
+function moveRequested(
```

Comment:
This feels like it should be named differently, something like `s3BucketKeyMatch` or something clever involving an identitiy property, or ...

Reply:
OBE